build/lib/android/snapshot.js

```diff
@@ -1,114 +1,75 @@
-const os = require('os');
 const fs = require('fs-extra');
 const path = require('path');
-const ejs = require('ejs');
+const request = require('request-promise-native');
 
 const ROOT_DIR = path.join(__dirname, '..', '..', '..');
 const DIST_DIR = path.join(ROOT_DIR, 'dist');
 const TMP_DIR = path.join(DIST_DIR, 'tmp');
 const ANDROID_DIR = path.join(ROOT_DIR, 'android');
 const ANDROID_PROPS = require(path.join(ANDROID_DIR, 'package.json')); // eslint-disable-line security/detect-non-literal-require
 const V8_PROPS = ANDROID_PROPS.v8;
-const V8_LIB_DIR = path.join(ROOT_DIR, 'dist', 'android', 'libv8', V8_PROPS.version, V8_PROPS.mode, 'libs');
-
-// Obtain target architectures
-const TARGETS = [];
-for (const target of ANDROID_PROPS.architectures) {
-	if (target === 'arm64-v8a') {
-		TARGETS.push('arm64');
-	} else if (target === 'armeabi-v7a') {
-		TARGETS.push('arm');
-	} else {
-		TARGETS.push(target);
-	}
-}
 
 /**
- * Runs mksnapshot to generate snapshots for each architecture
- * @param {string} target targets to generate blob
+ * Generates empty snapshot blobs for each supported architecture
+ * and creates a header to include the snapshots at build time.
  * @returns {Promise<void>}
  */
-async function generateBlob(target) {
-	const V8_LIB_TARGET_DIR = path.resolve(V8_LIB_DIR, target);
-	const MKSNAPSHOT_PATH = path.join(V8_LIB_TARGET_DIR, 'mksnapshot');
-	const BLOB_PATH = path.join(V8_LIB_TARGET_DIR, 'blob.bin');
-	const STARTUP_PATH = path.join(TMP_DIR, 'startup.js');
-	const TI_MAIN_PATH = path.join(ROOT_DIR, 'common', 'Resources', 'ti.main.js');
-	const TI_MAIN_PLATFORM_PATH = path.join(TMP_DIR, 'common', 'android', 'ti.main.js');
-	const args = [
-		'--startup_blob=' + BLOB_PATH,
-		STARTUP_PATH,
-		'--print-all-exceptions'
-	];
-
-	// Snapshot already exists, skip...
-	if (await fs.exists(BLOB_PATH)) {
-		const { blobStat, commonStat } = await Promise.all([ fs.stat(BLOB_PATH), fs.stat(TI_MAIN_PATH) ]);
-		if (commonStat && blobStat && commonStat.mtime < blobStat.mtime) {
-			return;
-		}
-	}
+async function build() {
+	const SNAPSHOT_URL = 'http://v8-snapshot.appcelerator.com';
+	const v8 = V8_PROPS.version;
+	const script = (await fs.readFile(path.join(TMP_DIR, 'common', 'android', 'ti.main.js'))).toString();
 
-	// Load platform optimized 'common' bundle
-	const commonBundle = await fs.readFile(TI_MAIN_PLATFORM_PATH);
+	return new Promise(async resolve => { // eslint-disable-line no-async-promise-executor
 
-	// Generate 'startup.js'
-	const output = await promisify(ejs.renderFile)(path.join(__dirname, 'startup.ejs'), { script: commonBundle }, {});
-	await fs.writeFile(STARTUP_PATH, output);
+		console.log('Attempting to request snapshot...');
 
-	// Set correct permissions for 'mksnapshot'
-	await fs.chmod(MKSNAPSHOT_PATH, 0o755);
+		// Obtain snapshot `id` and start new snapshot generation.
+		const id = await request.post(SNAPSHOT_URL, { body: { v8, script }, json: true });
```

Comment:
If you don't have Internet access, then the `await request.post()` will reject and throw an exception. Perhaps we should catch it, log the reject's error, and continue without the snapshot?

Although for Jenkins builds of our SDK (namely release builds), we don't want this behavior.

Reply:
We're catching the exception [here](https://github.com/appcelerator/titanium_mobile/blob/master/build/lib/android/index.js#L70). So this shouldn't be an issue.